auraed/src/vms/vm_service.rs

```diff
@@ -1,45 +1,129 @@
+use net_util::MacAddr;
 use proto::vms::{
-    vm_service_server, VirtualMachine, VmServiceCreateRequest,
-    VmServiceCreateResponse, VmServiceFreeRequest, VmServiceFreeResponse,
-    VmServiceStartRequest, VmServiceStartResponse, VmServiceStopRequest,
-    VmServiceStopResponse,
+    vm_service_server, VmServiceCreateRequest, VmServiceCreateResponse,
+    VmServiceFreeRequest, VmServiceFreeResponse, VmServiceStartRequest,
+    VmServiceStartResponse, VmServiceStopRequest, VmServiceStopResponse,
 };
-use std::sync::Arc;
+use std::{net::Ipv4Addr, path::PathBuf, sync::Arc};
 use tokio::sync::Mutex;
 use tonic::{Request, Response, Status};
+use vmm_sys_util::rand;
+
+use super::{
+    virtual_machine::{MountSpec, NetSpec, VmID, VmSpec},
+    virtual_machines::VirtualMachines,
+};
 
 #[derive(Debug, Clone)]
 pub struct VmService {
-    _vms: Arc<Mutex<VirtualMachine>>,
+    vms: Arc<Mutex<VirtualMachines>>,
+}
+
+impl VmService {
+    pub fn new() -> Self {
+        let vms = Arc::new(Mutex::new(VirtualMachines::new()));
+        Self { vms }
+    }
 }
 
 #[tonic::async_trait]
 impl vm_service_server::VmService for VmService {
     async fn create(
         &self,
-        _request: Request<VmServiceCreateRequest>,
+        request: Request<VmServiceCreateRequest>,
     ) -> Result<Response<VmServiceCreateResponse>, Status> {
-        todo!()
+        let mut vms = self.vms.lock().await;
+        let req = request.into_inner();
+
+        let Some(vm) = req.machine else {
+            return Err(Status::invalid_argument("No machine config provided"));
+        };
+
+        let Some(root_drive) = vm.root_drive else {
+            return Err(Status::invalid_argument("No root drive provided"));
+        };
+
+        let mut mounts = vec![MountSpec {
+            host_path: PathBuf::from(root_drive.host_path.as_str()),
+            read_only: !root_drive.is_writeable,
+        }];
+        mounts.extend(vm.drive_mounts.into_iter().map(|m| MountSpec {
+            host_path: PathBuf::from(m.host_path.as_str()),
+            read_only: !m.is_writeable,
+        }));
+
+        let net = vec![NetSpec {
+            tap: Some(format!(
+                "aurae0-{}",
+                rand::rand_alphanumerics(8).into_string().map_err(|_| {
+                    Status::internal("Failed to generate tap device name")
+                })?
+            )),
+            ip: Ipv4Addr::new(192, 168, 122, 1),
```

Comment:
TODO: Ipv6?
TODO: rotate the IP addresses on each call to create (if i'm reading this code right)?

Reply:
>rotate the IP addresses on each call to create (if i'm reading this code right)?
Yes, definitely should do that. I could start by implementing it naively and just iterate through a hard-coded subnet?

>Ipv6?
We can access over IPv6 link-local (`fe80::`) already, but it doesn't look like CH supports assigning ipv6 addresses to interfaces yet. I think we'd need set up the TAP manually to do this.

https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/net_util/src/open_tap.rs#L91

Reply:
> >rotate the IP addresses on each call to create (if i'm reading this code right)?
> 
> Yes, definitely should do that. I could start by implementing it naively and just iterate through a hard-coded subnet?

oh yeah that's fine.  254 VMs should be enough for anyone!

> 
> >Ipv6?
> 
> We can access over IPv6 link-local (`fe80::`) already, but it doesn't look like CH supports assigning ipv6 addresses to interfaces yet. I think we'd need set up the TAP manually to do this.

let's add a TODO.  I'd very much like to be ipv6 only by default all through the stack.  I'm a bit surprised CH hasn't made the same decision.

but it's all internal so doesn't matter that much.

> 
> https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/net_util/src/open_tap.rs#L91